analyses/doublet-detection/exploratory-analyses/02_explore-benchmark-results.Rmd

```diff
@@ -0,0 +1,267 @@
+  result_dir, 
+  glue::glue("{params$dataset}_scrublet.tsv")
+)
+truth_tsv <- file.path(
+  data_dir, 
+  glue::glue("{params$dataset}_ground-truth.tsv")
+)
+```
+
+### Functions
+
+This chunk defines functions to explore doublet detection results
+
+
+```{r functions}
+plot_doublets <- function(df, score_col, pred_col, method_name){
+  # df is expected to contain columns `ground_truth`, `score_col`, and `pred_col`
+  # The column arguments should _not_ be provided as strings
+  ggplot(df) +
+    aes(x = ground_truth,
+        y = {{score_col}}, 
+        color = {{pred_col}}) + 
+    geom_jitter(
+      alpha = 0.2, 
+      width = 0.1
+    ) + 
+    labs(
+      x = "Ground truth call",
+      y = glue::glue("{method_name} score"),
+      color = glue::glue("{method_name} prediction")
+    )
+}
+
+
+assess_doublets <- function(df, pred_col) {
+  # df is expected to contain columns `ground_truth` and `pred_col`
+  # The `pred_col` argument _should_ be provided as a strings
+  
+  caret::confusionMatrix(
+    # truth should be first
+    table(
+      "Truth" = doublet_df$ground_truth,
+      "Prediction" = doublet_df[[pred_col]]
+    ), 
+    positive = "doublet"
+  )
+  
+}
+
+```
+
+
+### Read and prepare input data
+
+First, we'll read in and combine TSV files with doublet results.
+
+```{r}
+scdbl_df <- read_tsv(scdbl_tsv) |>
+  dplyr::select(
+    barcodes,
+    cxds_score, 
+    scdbl_score = score, 
+    scdbl_prediction  = class
+  ) |>
+  # add cxds calls at three thresholds:
+  # 0.5, 0.75, 0.9
+  dplyr::mutate(
+    cxds_prediction_0.5 = dplyr::if_else(
+      cxds_score >= 0.5,
+      "doublet",
+      "singlet"
+    ),
+    cxds_prediction_0.75 = dplyr::if_else(
+      cxds_score >= 0.75,
+      "doublet",
+      "singlet"
+    ),
+    cxds_prediction_0.9 = dplyr::if_else(
+      cxds_score >= 0.9,
+      "doublet",
+      "singlet"
+    )
+  )
+
+scrub_df <- read_tsv(scrub_tsv) 
+
+true_df <- read_tsv(truth_tsv)
+
+doublet_df <- dplyr::inner_join(
+  scdbl_df,
+  scrub_df, 
+  by = "barcodes"
+) |>
+  dplyr::inner_join(
+    true_df, 
+    by = "barcodes"
+  )
+```
+
+## Explore results: Score distributions
+
+Both `scDblFinder` and `scrublet` make doublet calls based on an internally-determined threshold, which may not be optimal. 
+
+These sections explore the score distributions and how they relate to calls, including a confusion matrix and associated performance metrics. 
+We're particularly interested in "Balanced accuracy," as there is a much higher proportion of singlets vs doublets.
+
+### scDblFinder results
+
+```{r}
+plot_doublets(
+  doublet_df,
+  scdbl_score,
+  scdbl_prediction,
+  "scDblFinder"
+)
+
+assess_doublets(
+  doublet_df,
+  "scdbl_prediction"
+) |>
+  print()
+```
+
+
+### scrublet results
+
+```{r}
+plot_doublets(
+  doublet_df,
+  scrublet_score,
+  scrublet_prediction,
+  "scrublet"
+)
+
+assess_doublets(
+  doublet_df,
+  "scrublet_prediction"
+) |>
+  print()
+```
+
+
+### cxds results
```

Comment:
I think it would be helpful here to show the distributions of the score before doing any cutoffs? Maybe a density plot colored by ground truth so you can compare what scores look like for true singlets and true doublets.